mozetl/hardware_report/summarize_json.py

```diff
@@ -0,0 +1,716 @@
+
+        # Fetch the data we need.
+        data = frame.rdd.map(
+            lambda r: get_latest_valid_per_client(
+                r, chunk_start, chunk_end))
+
+        # Filter out broken data.
+        filtered_data = data.filter(
+            lambda r: r not in [
+                REASON_BROKEN_DATA,
+                REASON_INACTIVE])
+
+        # Count the broken records and inactive records.
+        discarded = data.filter(
+            lambda r: r in [
+                REASON_BROKEN_DATA,
+                REASON_INACTIVE]).countByValue()
+        broken_count = discarded[REASON_BROKEN_DATA]
+        inactive_count = discarded[REASON_INACTIVE]
+        broken_ratio = broken_count / float(records_count)
+        inactive_ratio = inactive_count / float(records_count)
+        print "Broken pings ratio: {}\nInactive clients ratio: {}".format(broken_ratio, inactive_ratio)
+
+        # If we're not seeing sane values for the broken or inactive ratios,
+        # bail out early on. There's no point in aggregating.
+        if broken_ratio >= 0.9 or inactive_ratio >= 0.9:
+            raise Exception(
+                "Unexpected ratio of broken pings or inactive clients.")
+
+        # Process the data, transforming it in the form we desire.
+        processed_data = filtered_data.map(prepare_data)
+
+        print "Aggregating entries..."
+        aggregated_pings = aggregate_data(processed_data)
+
+        # Get the sample count, we need it to compute the percentages instead of raw numbers.
+        # Since we're getting only the newest ping for each client, we can simply count the
+        # number of pings. THIS MAY NOT BE CONSTANT ACROSS WEEKS!
+        valid_records_count = filtered_data.count()
+
+        # Collapse together groups that count less than 1% of our samples.
+        threshold_to_collapse = int(valid_records_count * 0.01)
+
+        print "Collapsing smaller groups into the other bucket (threshold {th})".format(th=threshold_to_collapse)
+        collapsed_aggregates = collapse_buckets(
+            aggregated_pings, threshold_to_collapse)
+
+        print "Post-processing raw values..."
+
+        processed_aggregates = finalize_data(collapsed_aggregates,
+                                             valid_records_count,
+                                             broken_ratio,
+                                             inactive_ratio,
+                                             chunk_start)
+
+        if not validate_finalized_data(processed_aggregates):
+            raise Exception("The aggregates failed to validate.")
+
+        print "Serializing results locally..."
+        # This either appends to an existing file, or creates a new one.
+        serialize_results(processed_aggregates, chunk_start, chunk_end)
+
+        # Move on to the next chunk, just add one day the end of the last
+        # chunk.
+        chunk_start = chunk_end + dt.timedelta(days=1)
+
+
+@click.command()
+def main():
```

Comment:
Let's move the main function to a separate file. We should have it read in some CLI arguments (`start_date`, specifically, with `end_date` defaulted to one week after `start_date`), to make this easily compatible with airflow. Also add a file to the `scheduling` dir which we'll use to schedule this job, see the others in there for how it should be configured.